src/main/java/org/interledger/ilp/exceptions/InterledgerException.java

```diff
@@ -1,36 +1,196 @@
+        F03_INVALID_AMOUNT            ("F03", "INVALID AMOUNT"          ),
+        F04_INSUFFICIENT_DST_AMOUNT   ("F04", "INSUFFICIENT DST. AMOUNT"),
+        // TODO:(RFC) WRONG CONDITION is actually WRONG FULFILLMENT?
+        F05_WRONG_CONDITION           ("F05", "WRONG CONDITION"         ),
+        F06_UNEXPECTED_PAYMENT        ("F06", "UNEXPECTED PAYMENT"      ),
+        F07_CANNOT_RECEIVE            ("F07", "CANNOT RECEIVE"          ),
+        F99_APPLICATION_ERROR         ("F99", "APPLICATION ERROR"       ),
+
+        T00_INTERNAL_ERROR            ("T00", "INTERNAL ERROR"          ),
+        T01_LEDGER_UNREACHABLE        ("T01", "LEDGER UNREACHABLE"      ),
+        T02_LEDGER_BUSY               ("T02", "LEDGER BUSY"             ),
+        T03_CONNECTOR_BUSY            ("T03", "CONNECTOR BUSY"          ),
+        T04_INSUFFICIENT_LIQUIDITY    ("T04", "INSUFFICIENT LIQUIDITY"  ),
+        T05_RATE_LIMITED              ("T05", "RATE LIMITED"            ),
+        T99_APPLICATION_ERROR         ("T99", "APPLICATION ERROR"       ),
+
+        R00_TRANSFER_TIMED_OUT        ("R00", "TRANSFER TIMED OUT"      ),
+        R01_INSUFFICEINT_SOURCE_AMOUNT("R01", "INSUFFICEINT SOURCE AMOUNT"),
+        R02_INSUFFICIENT_TIMEOUT      ("R02", "INSUFFICIENT TIMEOUT"    ),
+        R99_APPLICATION_ERROR         ("R99", "APPLICATION ERROR"       )
+        ;
+        // TODO:(RFC) ADD FORBIDEN ERROR IF NO CREDENTIALS ARE PROVIDED OR LAUNCH "T01 LEDGER-UNREACHABLE" ???
+        // IMAGINE node1 -> conector1 -> conector2 -> (ERR. FORBIDEN) contector3
+        //  - How must routing be updated in  conector1/2/3 ?
+
+        public  final ErrorType errorType;
+        private final String code;
+        private final String name;
+
+        private ErrorCode(String code, String name) {
+            code = Objects.requireNonNull(code, "code must not be null").trim();
+            name = Objects.requireNonNull(name, "code must not be null").trim();
+
+            if (code.length()!=3) {
+                throw new RuntimeException("error code length must be equal to 3");
+            }
+            char type = code.charAt(0);
+            switch (type) {
+            case 'F':
+                errorType = ErrorType.FINAL;
+                break;
+            case 'T':
+                errorType = ErrorType.TEMPORARY;
+                break;
+            case 'R':
+                errorType = ErrorType.RELATIVE;
+                break;
+            default:
+                throw new IllegalArgumentException("code must start with 'F', 'T' or 'R'.");
+            }
+
+            this.code = code;
+            this.name = name;
+        }
+
+        @Override
+        public String toString(){
+            return this.code + "-" +this.name;
+        }
+    }
+
+    private final ErrorCode errCode;
+    private final InterledgerAddress triggeredBy;
+
+    private final InterledgerAddress[] forwardedBy;
```

Comment:
I prefer not using Arrays for a variety of reasons.  See [here](http://stackoverflow.com/questions/6100148/collection-interface-vs-arrays) or [here](http://eclipsesource.com/blogs/2014/04/11/3-good-reasons-to-avoid-arrays-in-java-interfaces/) for some of them.  If you want to use an Array internally, I would be ok with that, but our interface (i.e., the constructor, etc) should use a Collection.

Thoughts?